File: internal/parity/runner.go
package parity

import (
	"bytes"
	"fmt"
	"strings"
)

// Options controls normalization and diagnostic collection.
type Options struct {
	Normalizer Normalizer // CanonicalJSON when nil.
	MaxDiffs   int        // 20 when zero.
}

// Diff describes one mismatched or errored fixture.
type Diff struct {
	ID       string
	Expected string
	Got      string
	Err      string
}

// Report summarizes a differential harness run.
type Report struct {
	Total      int
	Ran        int
	Matched    int
	Mismatched int
	Errored    int
	Diffs      []Diff
}

// OK reports whether every fixture run matched its expected value.
func (r Report) OK() bool {
	return r.Mismatched == 0 && r.Errored == 0
}

// String returns a human-readable summary followed by the collected diffs.
func (r Report) String() string {
	var out strings.Builder
	fmt.Fprintf(
		&out,
		"total=%d ran=%d matched=%d mismatched=%d errored=%d",
		r.Total,
		r.Ran,
		r.Matched,
		r.Mismatched,
		r.Errored,
	)

	for index, diff := range r.Diffs {
		fmt.Fprintf(&out, "\n\ndiff %d: id=%q", index+1, diff.ID)
		if diff.Err != "" {
			fmt.Fprintf(&out, "\n  error: %s", diff.Err)
		}
		if diff.Expected != "" {
			fmt.Fprintf(&out, "\n  expected: %s", diff.Expected)
		}
		if diff.Got != "" {
			fmt.Fprintf(&out, "\n  got: %s", diff.Got)
		}
	}

	omitted := r.Mismatched + r.Errored - len(r.Diffs)
	if omitted > 0 {
		fmt.Fprintf(&out, "\n\n... %d additional diffs omitted", omitted)
	}

	return out.String()
}

// Run applies d to fixtures for its subsystem and compares normalized output
// with each fixture's normalized expected value.
func Run(fixtures []Fixture, d Driver, opts Options) Report {
	normalizer := opts.Normalizer
	if normalizer == nil {
		normalizer = CanonicalJSON
	}

	maxDiffs := opts.MaxDiffs
	if maxDiffs == 0 {
		maxDiffs = 20
	}

	report := Report{Total: len(fixtures)}
	appendDiff := func(diff Diff) {
		if len(report.Diffs) < maxDiffs {
			report.Diffs = append(report.Diffs, diff)
		}
	}

	subsystem := d.Subsystem()
	for _, fixture := range fixtures {
		if fixture.Subsystem != subsystem {
			continue
		}

		report.Ran++
		got, err := d.Run(fixture.Input)
		if err != nil {
			report.Errored++
			appendDiff(Diff{ID: fixture.ID, Err: err.Error()})
			continue
		}

		normalizedGot, err := normalizer(got)
		if err != nil {
			report.Errored++
			appendDiff(Diff{
				ID:       fixture.ID,
				Expected: string(fixture.Expected),
				Got:      string(got),
				Err:      fmt.Sprintf("normalize driver output: %v", err),
			})
			continue
		}

		normalizedExpected, err := normalizer(fixture.Expected)
		if err != nil {
			report.Errored++
			appendDiff(Diff{
				ID:       fixture.ID,
				Expected: string(fixture.Expected),
				Got:      string(normalizedGot),
				Err:      fmt.Sprintf("normalize expected value: %v", err),
			})
			continue
		}

		if bytes.Equal(normalizedGot, normalizedExpected) {
			report.Matched++
			continue
		}

		report.Mismatched++
		appendDiff(Diff{
			ID:       fixture.ID,
			Expected: string(normalizedExpected),
			Got:      string(normalizedGot),
		})
	}

	return report
}

File: bitmagnet-rs/crates/bitmagnet-diff/src/runner.rs
use std::fmt;

use crate::driver::Driver;
use crate::fixture::Fixture;
use crate::normalize::{canonical, Normalizer};

/// Controls normalization and diagnostic collection.
pub struct Options {
    /// Function used to canonicalize actual and expected JSON values.
    pub normalizer: Normalizer,
    /// Maximum number of errored or mismatched fixtures retained in `diffs`.
    pub max_diffs: usize,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            normalizer: canonical,
            max_diffs: 20,
        }
    }
}

/// One mismatched or errored fixture.
#[derive(Debug, Clone)]
pub struct Diff {
    pub id: String,
    pub expected: String,
    pub got: String,
    pub err: Option<String>,
}

/// Summary of a differential harness run.
#[derive(Debug, Default)]
pub struct Report {
    pub total: usize,
    pub ran: usize,
    pub matched: usize,
    pub mismatched: usize,
    pub errored: usize,
    pub diffs: Vec<Diff>,
}

impl Report {
    /// Whether every fixture run matched its expected value.
    #[must_use]
    pub fn ok(&self) -> bool {
        self.mismatched == 0 && self.errored == 0
    }
}

impl fmt::Display for Report {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "total={} ran={} matched={} mismatched={} errored={}",
            self.total, self.ran, self.matched, self.mismatched, self.errored
        )?;

        for (index, diff) in self.diffs.iter().enumerate() {
            write!(formatter, "\n\ndiff {}: id={:?}", index + 1, diff.id)?;
            if let Some(err) = &diff.err {
                write!(formatter, "\n  error: {err}")?;
            }
            if !diff.expected.is_empty() {
                write!(formatter, "\n  expected: {}", diff.expected)?;
            }
            if !diff.got.is_empty() {
                write!(formatter, "\n  got: {}", diff.got)?;
            }
        }

        let failed = self.mismatched.saturating_add(self.errored);
        let omitted = failed.saturating_sub(self.diffs.len());
        if omitted > 0 {
            write!(formatter, "\n\n... {omitted} additional diffs omitted")?;
        }

        Ok(())
    }
}

/// Run a driver against every fixture for its subsystem.
#[must_use]
pub fn run<D: Driver>(fixtures: &[Fixture], driver: &D, opts: Options) -> Report {
    let mut report = Report {
        total: fixtures.len(),
        ..Report::default()
    };

    let subsystem = driver.subsystem();
    for fixture in fixtures {
        if fixture.subsystem != subsystem {
            continue;
        }

        report.ran += 1;
        let got = match driver.run(&fixture.input) {
            Ok(got) => got,
            Err(err) => {
                report.errored += 1;
                if report.diffs.len() < opts.max_diffs {
                    report.diffs.push(Diff {
                        id: fixture.id.clone(),
                        expected: String::new(),
                        got: String::new(),
                        err: Some(err.to_string()),
                    });
                }
                continue;
            }
        };

        let normalized_got = (opts.normalizer)(&got);
        let normalized_expected = (opts.normalizer)(&fixture.expected);
        if normalized_got == normalized_expected {
            report.matched += 1;
            continue;
        }

        report.mismatched += 1;
        if report.diffs.len() < opts.max_diffs {
            report.diffs.push(Diff {
                id: fixture.id.clone(),
                expected: normalized_expected.to_string(),
                got: normalized_got.to_string(),
                err: None,
            });
        }
    }

    report
}
